backends/cloud/box/boxstorage.cpp

```diff
@@ -0,0 +1,299 @@
+	debug("%s", error.response.c_str());
+	CloudMan.removeStorage(this);
+}
+
+void BoxStorage::saveConfig(Common::String keyPrefix) {
+	ConfMan.set(keyPrefix + "access_token", _token, ConfMan.kCloudDomain);
+	ConfMan.set(keyPrefix + "refresh_token", _refreshToken, ConfMan.kCloudDomain);
+}
+
+Common::String BoxStorage::name() const {
+	return "Box";
+}
+
+void BoxStorage::infoInnerCallback(StorageInfoCallback outerCallback, Networking::JsonResponse response) {
+	Common::JSONValue *json = response.value;
+	if (!json) {
+		warning("NULL passed instead of JSON");
+		delete outerCallback;
+		return;
+	}
+
+	Common::JSONObject info = json->asObject();
+
+	Common::String uid, name, email;
+	uint64 quotaUsed = 0, quotaAllocated = 0;
+
+	// can check that "type": "user"
+	// there is also "max_upload_size", "phone" and "avatar_url"
+
+	if (info.contains("id") && info.getVal("id")->isString())
+		uid = info.getVal("id")->asString();
+
+	if (info.contains("name") && info.getVal("name")->isString())
+		name = info.getVal("name")->asString();
+
+	if (info.contains("login") && info.getVal("login")->isString())
+		email = info.getVal("login")->asString();
+
+	if (info.contains("space_amount") && info.getVal("space_amount")->isIntegerNumber())
+		quotaAllocated = info.getVal("space_amount")->asIntegerNumber();
+
+	if (info.contains("space_used") && info.getVal("space_used")->isIntegerNumber())
+		quotaUsed = info.getVal("space_used")->asIntegerNumber();
+
+	Common::String username = email;
+	if (username == "") username = name;
+	if (username == "") username = uid;
+	CloudMan.setStorageUsername(kStorageBoxId, username);
+
+	if (outerCallback) {
+		(*outerCallback)(StorageInfoResponse(nullptr, StorageInfo(uid, name, email, quotaUsed, quotaAllocated)));
+		delete outerCallback;
+	}
+
+	delete json;
+}
+
+Networking::Request *BoxStorage::listDirectoryById(Common::String id, ListDirectoryCallback callback, Networking::ErrorCallback errorCallback) {
+	if (!errorCallback) errorCallback = getErrorPrintingCallback();
+	if (!callback) callback = getPrintFilesCallback();
+	return addRequest(new BoxListDirectoryByIdRequest(this, id, callback, errorCallback));
+}
+
+void BoxStorage::createDirectoryInnerCallback(BoolCallback outerCallback, Networking::JsonResponse response) {
+	Common::JSONValue *json = response.value;
+	if (!json) {
+		warning("NULL passed instead of JSON");
+		delete outerCallback;
+		return;
+	}
+
+	if (outerCallback) {
+		Common::JSONObject info = json->asObject();
+		(*outerCallback)(BoolResponse(nullptr, info.contains("id")));
+		delete outerCallback;
+	}
+
+	delete json;
+}
+
+Networking::Request *BoxStorage::createDirectoryWithParentId(Common::String parentId, Common::String name, BoolCallback callback, Networking::ErrorCallback errorCallback) {
+	if (!errorCallback) errorCallback = getErrorPrintingCallback();
+
+	Common::String url = "https://api.box.com/2.0/folders";
```

Comment:
I recommend to move these to the very top of the file as constants with #define.